#4 test integration

XioZ opened the pull request:
- Implemented `Mino` and `RPC` interfaces to make calls/streams
- Closing #4

XioZ committed:
#3 Added unit tests for address.go

XioZ committed:
#3 Uses table tests and subtests

XioZ committed:
#3 Cleaned up TODOs

XioZ committed:
#4 Added Minows and RPC. Updated Address and tests

XioZ committed:
#4 Added integration test for Call()

XioZ committed:
#4 Implementing Call()

XioZ committed:
#4 Implementing Stream()

XioZ committed:
#5 Starts CLI controller

XioZ committed:
#5 Refactored Address tests

XioZ committed:
#5 Refactored Address tests

XioZ committed:
#4 Implemented Stream()

XioZ committed:
#4 Implemented Stream()

XioZ committed:
#4 Wrap open stream result with address and error

XioZ committed:
#4 Refactored Stream()

XioZ committed:
#4 Refactored Call()

XioZ committed:
Merge branch 'feature/#5-add-cli' into feature/#4-test-integration

# Conflicts:
#	mino/minows/controller/mod.go
#	mino/minows/rpc.go
#	mino/minows/session.go

XioZ committed:
#4 Resolved todos in session.go

XioZ committed:
#4 Resolved todos in address.go

XioZ committed:
#4 Resolved todos in mod.go

XioZ committed:
#4 Resolved todos in rpc.go

ineiti commented on go.mod:
```diff
@@ -1,67 +1,117 @@
 module go.dedis.ch/dela
 
-go 1.19
+go 1.21
```
Do you need a new version for libp2p? If not, revert these changes and stay with 1.19 - changing the go.mod of a big project is not an easy thing to do. E.g., 1.22 breaks some of the tests...

XioZ replied:
didn't realize the go version got updated automatically when i ran `go get libp2p`. reverted `go.mod` file and using go-libp2p v0.29 now (instead of 0.33) to be compatible with go v1.19. tests still pass

ineiti commented on mino/minows/address.go:
```diff
@@ -0,0 +1,93 @@
+// ConnectionType implements mino.Address
+func (a address) ConnectionType() mino.AddressConnectionType {
+	// TODO implement
+	panic("not implemented")
```
This looks easy to me to implement - is there something I'm missing?

XioZ replied:
adding a new `AddressConnectionType` const previously broke some tests in grpc, so left it last to implement. returning `ACTws` now but this method is not used.

ineiti commented on mino/minows/address_test.go:
```diff
@@ -0,0 +1,232 @@
+)
+
+const (
+	AddrAllInterface = "/ip4/0.0.0.0/tcp/80"
+	AddrLocalhost    = "/ip4/127.0.0.1/tcp/80"
+	AddrHostname     = "/dns4/example.com/tcp/80"
```
Please prepend all of these with `p2p://`, so it's `p2p://ipv4/0.0.0.0/tcp/80`. This will make it more similar to an URL.

XioZ replied:
multiaddress requires a certain format. [this doc](https://docs.libp2p.io/concepts/fundamentals/addressing/) contains more details.

ineiti commented on mino/minows/address_test.go:
```diff
@@ -0,0 +1,232 @@
+		},
+	}
+
+	t.Parallel()
```
From what I understand reading https://pkg.go.dev/testing#T.Parallel, this is not needed. Only the one on line 201.

XioZ replied:
first call parallelizes the top-level test function to other test functions. second parallelizes the subtest to other subtests in the table test. [this doc](https://go.dev/wiki/TableDrivenTests#parallel-testing) contains more details.

ineiti replied:
See my other comments - why add `Parallel` at all here? It's nothing taking a long time, and no other package in dela uses `Parallel`. So the confusion / usefulness ratio is too high in my opinion.

Remove `Parallel` for your tests, unless you have a good reason to have it here.

ineiti commented on mino/minows/integration_test.go:
```diff
@@ -0,0 +1,58 @@
+package minows
+
+//
+// import (
+// 	"context"
+// 	"testing"
+// 	"time"
+//
+// 	"github.com/stretchr/testify/require"
```
Remove commented out code.

XioZ replied:
removed file

ineiti commented on mino/minows/mod.go:
```diff
@@ -0,0 +1,104 @@
+	context  serde.Context
+}
+
+// NewMinows
```
Add a short one-sentence description.

XioZ replied:
auto generated doc last. please check updated comments

ineiti commented on mino/minows/rpc.go:
```diff
@@ -0,0 +1,235 @@
+// - When it's done, all the connections are shut down and the resources are
+// cleaned up (DELA Doc)
+// - orchestrator of a protocol will contact one of the participants which
+// will be the root for the routing algorithm (i.e. gateway?).
```
```suggestion
// will be the root for the routing algorithm (the orchestrator).
```

ineiti commented on mino/minows/rpc.go:
```diff
@@ -0,0 +1,235 @@
+// cleaned up (DELA Doc)
+// - orchestrator of a protocol will contact one of the participants which
+// will be the root for the routing algorithm (i.e. gateway?).
+// It will then relay the messages according to the routing algorithm and
+// create relays to other peers when necessary (DELA Doc but ignored)
```
```suggestion
// It will then relay the messages according to the routing algorithm, which is
// a direct connection between every pair of nodes in our case.
```

ineiti reviewed (changes requested):
Starts looking good. Needs some changes.

Also there are a lot of TODOs in the code - do you have any idea if all of them are necessary for the final implementation, or some of them are optional?

XioZ committed:
#4 Handles Call() & Stream()

XioZ committed:
#4 Added tests for minows

XioZ committed:
#4 Adding tests for rpc.go & session.go

XioZ committed:
#4 Added tests for rpc.go & session.go

ineiti commented on mino/minows/address.go:
```diff
@@ -86,7 +86,7 @@ func (f addressFactory) FromText(text []byte) mino.Address {
 	addr, err := newAddress(location, identity)
 	if err != nil {
-		dela.Logger.Err(xerrors.Errorf("could not create address: %v", err))
+		dela.Logger.Err(xerrors.Errorf("could not create address: %w", err))
```
Please revert. The standard error-passing mechanism in dela is using `%v`.

XioZ replied:
changed, though i'm not sure how this makes error assertion `errors.Is()` still possible at the caller.

ineiti replied:
Not sure, either. But sometimes it's more important to use common style than to support a feature nobody uses.

If you think that `errors.Is()` is not possible, open an issue on the dedis/dela repository, and express your concerns.

ineiti commented on mino/minows/address_test.go:
```diff
@@ -9,65 +9,78 @@ import (
+			args: args{addrHostname, pid1},
 		},
 	}
+	t.Parallel() // run this test function args parallel to other test functions
```
Here and elsewhere: the dela guid asks comments to be on their own line. So this should be:
```suggestion
    // run this test function args parallel to other test functions
	t.Parallel() 
```

And commenting obvious things should be avoided, like this line.

XioZ replied:
moved to own lines. maybe it's not obvious at first glance, since the two parallel calls are parallelizing diff things?

ineiti replied:
Yes, true. I was confused by the fact that tests from multiple packages are run in parallel, but the tests within one package are run sequentially:

https://bryce.is/writing/code/go-test-and-parallelism

(Which also explains the PAUSE we saw in one of the test runs)

Again, for the unity of the code, I would not run the tests in parallel here. Or only if you have a really good reason, like running the tests 10'000 times or so.

I try to avoid surprises like these, for debugging's sake. As all the other tests in `dela` are sequential, please make these tests also sequential.

ineiti commented on mino/minows/address_test.go:
```diff
@@ -9,65 +9,78 @@ import (
+			location: mustCreateMultiaddress(t, addrAllInterface),
+			identity: "",
+		}}
 	t.Parallel() // run this test function in parallel to other test functions
```
Aren't all tests run in parallel anyway? Except the ones where you call `t.Run`, where `t.Parallel` makes sense.

XioZ replied:
from [this doc](https://go.dev/wiki/TableDrivenTests#parallel-testing) it looks like by default `t.Run()` runs test functions sequentially? correct me if i misunderstood.

ineiti replied:
You're correct. It's tests from different packages which are run in parallel. See my comment above.

ineiti commented on mino/minows/address_test.go:
```diff
@@ -9,65 +9,78 @@ import (
+func Test_newAddress(t *testing.T) {
+	const pid1 = "QmaD31nEzFGwD8dK96UFWHtTYTqYJgHLMYSFz4W4Hm2WCU"
+	const pid2 = "QmVt9t5Tk2uEoA4CDbKNCVxqrut8UXmWHXvFZ8wFZ3ghhv"
+	const addrAllInterface = "/ip4/0.0.0.0/tcp/80"
+	const addrLocalhost = "/ip4/127.0.0.1/tcp/80"
+	const addrHostname = "/dns4/example.com/tcp/80"
```
Please add some protocol identifier like `p2pws://ip4/0.0.0.0/tcp/80`

XioZ replied:
`ip4` and `tcp` are the predefined protocol identifiers according to [this doc ](https://pkg.go.dev/github.com/multiformats/go-multiaddr#readme-example)

ineiti replied:
I'm more concerned with mino-identifiers. Like the `grpc://` and `grpc://` ones used for minogrpc.

On second thought I'm not sure it makes that much sense, as one dela-cli can only handle one mino-implementation anyway. Let's leave it like that.

ineiti commented on mino/minows/integration_test.go:
```diff
@@ -45,7 +45,7 @@ package minows
 //
 // func newMinowsAndRPC() (mino.Mino, mino.RPC) {
-// 	m, err := NewMinows("ip4/127.0.0.1/tcp/9998/ws", "ip4/127.0.0.1/tcp/9998/ws",
+// 	m, err := newMinows("ip4/127.0.0.1/tcp/9998/ws", "ip4/127.0.0.1/tcp/9998/ws",
```
For the whole file: please don't commit commented out code. Remove the commented code from the commit.

XioZ replied:
removed file

ineiti commented on mino/minows/mod.go:
```diff
@@ -17,40 +20,43 @@ import (
-// identity: Peer ID (multi-hash of public key generated by libp2p)
-func NewMinows(listen, public ma.Multiaddr, privKey crypto.PrivKey) (*minows, error) {
-	id, err := peer.IDFromPrivateKey(privKey)
+// newMinows
```
There should be a phrase at the beginning, e.g.,

```
// newMinows starts a new mino service using libp2p to connect over websockets.
```

XioZ replied:
generated doc now

ineiti commented on mino/minows/mod.go:
```diff
@@ -17,40 +20,43 @@ import (
+	id, err := peer.IDFromPrivateKey(secret)
 	if err != nil {
-		return nil, xerrors.Errorf("could not get Peer ID: %v", err)
+		return nil, xerrors.Errorf("could not get Peer ID: %w", err)
```
Here and elsewhere: please use `%v`

XioZ replied:
changed

ineiti commented on mino/minows/mod.go:
```diff
@@ -87,18 +97,74 @@ func (m *minows) CreateRPC(name string, h mino.Handler, f serde.Factory) (mino.R
+	m.host.SetStreamHandler(protocol.ID(uri+PostfixStream),
+		r.createStreamHandler(h))
+	// TODO when to return pointer vs struct?
+	return r, nil
```
https://bryanftan.medium.com/accept-interfaces-return-structs-in-go-d4cab29a301b

But I'm not sure how dela handles this in other modules.

XioZ replied:
returns pointer now

ineiti reviewed (changes requested):
Looks good - there are some changes requested - please change them, if possible before our meeting.

ineiti commented:
- Needs to be rebased on the latest `main`
- Please include my comments

XioZ committed:
#4 Refactoring

XioZ committed:
#4 Reverted go.mod

XioZ committed:
#4 Added go-libp2p as dependency in go.mod

XioZ committed:
#4 Resolved todos

XioZ committed:
#4 Fixed tests

XioZ committed:
#4 Fixed tests

XioZ committed:
#4 Fixed tests

XioZ committed:
#4 Fixed tests

XioZ committed:
#4 Fixed tests

XioZ commented:
just to clarify, the todos and comments were mostly for my own reference during development, and the pr was marked as draft to indicate not ready for review. i've resolved the todos now (only 1 left on `MaxMessageSize`) and generated doc so now it should be in shape for review.

XioZ commented on mino/minows/rpc.go:
```diff
@@ -0,0 +1,318 @@
+	"golang.org/x/xerrors"
+)
+
+const MaxMessageSize = 1e9 // TODO verify
```
copied this from grpc as the buffer size to use when reading from byte stream. please confirm it's ok?

XioZ committed:
#4 Updated doc

ineiti commented:
> just to clarify, the todos and comments were mostly for my own reference during development, and the pr was marked as draft to indicate not ready for review. i've resolved the todos now (only 1 left on `MaxMessageSize`) and generated doc so now it should be in shape for review.

OK, point taken. You did request a review from me, though. This happens if you choose one of the collaborators as `Reviewer`. So you should:

- start in draft
- work on it
- put it in 'ready for merge'
- request a review - if possible within the first half of the week

ineiti commented on mino/minows/mod_test.go:
```diff
@@ -0,0 +1,212 @@
+	var tests = map[string]struct {
+		args args
+	}{
+		// 'public' only uses localhost a for local testing
```
What do you mean by that?

ineiti commented on mino/minows/rpc.go:
```diff
@@ -0,0 +1,320 @@
+// Note:
+// - 'ctx' defines when the stream session ends,
+// so should always be canceled at some point (e.g. completed task or errored).
+// - When it's done, all the connections are shut down and resources freed.
```
This is not clear to me. `ctx` MUST cancel to close down?
When is `done`? When `ctx` is canceled?

XioZ replied:
a [context is done](https://pkg.go.dev/context#Context) when cancelled or times out (depending on how it's created), or its parent context is done. i copied these 2 bullet points from [dela doc](https://dedis.github.io/dela/#/mino?id=stream-bidirectional-stream-based-protocol). please let me know how it can be rephrased for clarity

ineiti commented on mino/minows/rpc.go:
```diff
@@ -0,0 +1,320 @@
+	if players == nil || players.Len() == 0 {
+		return nil, nil, xerrors.New("no players to stream")
+	}
+	// quit unless all player addresses are valid
```
Avoid negations in comments, it takes more time to read. Better:
```
// quit if any address is invalid
```

or

```
// Ensure all addresses are valid
```

or add this to the comment at the top of the method. If somebody reads these lines, they will already understand what's going on. But somebody might only read the documentation.

ineiti commented on mino/minows/mod.go:
```diff
@@ -0,0 +1,125 @@
+}
+
+func (m *minows) CreateRPC(name string, h mino.Handler, f serde.Factory) (mino.RPC, error) {
+	pattern := regexp.MustCompile("^[a-zA-Z0-9]+$")
```
You could optimize it a bit by adding this into a package-var, like here:

https://github.com/dedis/dela/blob/master/mino/minogrpc/mod.go#L417

XioZ replied:
changed

ineiti commented on mino/minows/mod.go:
```diff
@@ -0,0 +1,125 @@
+func (m *minows) CreateRPC(name string, h mino.Handler, f serde.Factory) (mino.RPC, error) {
+	pattern := regexp.MustCompile("^[a-zA-Z0-9]+$")
+	// validate namespace once
+	if len(m.rpcs) == 0 {
```
I wonder whether we should put that check in the `minows.WithSegment` method. The only downside is that `WithSegment` doesn't return an error, and you'd have to panic.

But having an invalid segment name would be a valid reason for me to panic.

ineiti commented on mino/minows/mod.go:
```diff
@@ -0,0 +1,125 @@
+		r.createCallHandler(h))
+	m.host.SetStreamHandler(protocol.ID(uri+PostfixStream),
+		r.createStreamHandler(h))
+	m.rpcs[name] = nil
```
Why do you delete the rpc entry here? Now you can create twice the same RPC name, no?

XioZ replied:
thereafter, `_, found := m.rpcs[name]` would return `found` as true. please see `Test_minows_CreateRPC_AlreadyExists` for verified behavior.

ineiti reviewed (commented):
- Please add some empty lines in your methods, according to https://dedis.github.io/dela/#/guideline
- avoid useless comments, like `form uri` or `create rpc` in the `minows.CreateRPC` method. This only catches the attention of the reader and doesn't help.
- still some comments on the same line as the code, please put on their own line
- comments should be a complete sentence (subject, verb), start with a capital letter, and end with a `.`
- rebase on the latest main
- please respect my week-end and don't ask me to review on Friday afternoons!

XioZ committed:
Merge remote-tracking branch 'origin/main' into feature/#4-test-integration

# Conflicts:
#	mino/minows/README.md

XioZ commented:
> > just to clarify, the todos and comments were mostly for my own reference during development, and the pr was marked as draft to indicate not ready for review. i've resolved the todos now (only 1 left on `MaxMessageSize`) and generated doc so now it should be in shape for review.
> 
> OK, point taken. You did request a review from me, though. This happens if you choose one of the collaborators as `Reviewer`. So you should:
> 
> * start in draft
> * work on it
> * put it in 'ready for merge'
> * request a review - if possible within the first half of the week

i think there's a miscommunication here. 2 meetings ago, you asked me to open a new PR whenever i start working on an issue and add you as a reviewer so you see my progress. so i did that with the draft PR and pushed my commits occasionally. i marked this PR ready for review today after seeing your comments. i don't mean to pressure you to review it right away, sorry. please do it whenever is convenient and i'll keep working on the next issue meanwhile. i'll address your comments as they come along. have a good weekend.

XioZ committed:
#5 Fixed comments

XioZ committed:
#5 Fixed comments

XioZ committed:
#5 Fixing lint

XioZ committed:
#5 Fixing tests

XioZ committed:
#5 Fixed lint

XioZ committed:
#5 Fixing tests

XioZ committed:
#5 Fixing tests

ineiti commented on mino/minows/rpc.go:
```diff
@@ -0,0 +1,332 @@
+// Note:
+// - 'ctx' defines when the stream session ends,
+// so should always be canceled at some point (e.g. completed task or errored).
+// - When it's done, all the connections are shut down and resources freed.
```
```suggestion
// Note: to end the session, the 'ctx' needs to get into 'done' state by cancelling, raising an error, or completing the task. When the session ends, all the connections are shut down and resources freed.
```

XioZ committed:
#4 Reads msg from stream using a growing buffer

XioZ committed:
#4 Read/Write network stream using json.Decoder/Encoder

XioZ commented:
Addressed PR comments. Uses a JSON encoder/decoder to write to/read from libp2p streams (instead of manually allocating a byte buffer each time) to address memory concerns.

XioZ committed:
#4 Uses mino.Response to channel received messages for session

XioZ committed:
Revert "#4 Uses mino.Response to channel received messages for session"

This reverts commit 1800a40c00c6b121b4a8f2e31d9f121c66f4d0d9.

XioZ committed:
#4 Uses envelope for session send & receive
